src/iblt.h

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright (c) 2018 The Bitcoin Unlimited developers
+Copyright (c) 2014 Gavin Andresen
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+#ifndef CIblt_H
+#define CIblt_H
+
+#include "serialize.h"
+
+#include <inttypes.h>
+#include <set>
+#include <vector>
+
+//
+// Invertible Bloom Lookup Table implementation
+// References:
+//
+// "What's the Difference? Efficient Set Reconciliation
+// without Prior Context" by Eppstein, Goodrich, Uyeda and
+// Varghese
+//
+// "Invertible Bloom Lookup Tables" by Goodrich and
+// Mitzenmacher
+//
+
+class HashTableEntry
+{
+public:
+    int32_t count;
+    uint64_t keySum;
+    uint32_t keyCheck;
+    std::vector<uint8_t> valueSum;
+
+    bool isPure() const;
+    bool empty() const;
+    void addValue(const std::vector<uint8_t> v);
+
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream &s, Operation ser_action)
+    {
+        READWRITE(count);
+        READWRITE(keySum);
+        READWRITE(keyCheck);
+        READWRITE(valueSum);
+    }
+};
+
+class CIblt
+{
+public:
+    CIblt();
+    CIblt(size_t _expectedNumEntries, size_t _ValueSize);
+    CIblt(const CIblt &other);
+    virtual ~CIblt();
+
+    void reset();
+    size_t size();
+    void resize(size_t _expectedNumEntries, size_t _ValueSize);
+    void insert(uint64_t k, const std::vector<uint8_t> v);
+    void erase(uint64_t k, const std::vector<uint8_t> v);
```

Comment:
pass the vectors by reference to avoid making a copy
void insert(uint64_t k, const std::vector<uint8_t> &v);
void erase(uint64_t k, const std::vector<uint8_t> &v);